include/array/array.h

```diff
@@ -2264,12 +2264,39 @@ class array {
    * `alloc`. This is a deep copy of the contents of `other`. */
   array(const array& other, const Alloc& alloc) : array(alloc) { assign(other); }
 
+  /** Move construct from another array `other`. If the allocator of this array
+   * is propagate on container move assignment, this operation moves the allocation
+   * of other to this array, and the other array becomes a default constructed
+   * array. If the allocator of this and the other array are non-equal, each
+   * element is move-constructed into a new allocation. */
+  array(array&& other) : buffer_(nullptr), buffer_size_(0), base_(nullptr) {
+    using std::swap;
+    if (std::allocator_traits<allocator_type>::propagate_on_container_move_assignment::value) {
```

Comment:
I think this really is diverging from what the C++ standard says: https://en.cppreference.com/w/cpp/named_req/AllocatorAwareContainer

> Move constructors obtain their instances of allocators by move-constructing from the allocator belonging to the old container.

I see now that what you have here is basically identical to the move assignment operator we already have, as you might expect after reading the above link. But the above also makes it seem like move construction should be different.

However, it does seem like an oversight of the C++ standard, because following that means it's not possible to fix the issue you are trying to fix (AFAICT). That, or the C++ standard implicitly outlaws `auto_allocator`-like things.

Reply:
> I see now that what you have here is basically identical to the move assignment operator we already have

I don't think so. There is `swap(alloc_, other.alloc_);`, which is exactly move constructing from the allocator belonging to the old container.

It is just the else case that is violating the C++ standard. The else case didn't get worse because as you said, it was identical to the move constructor with the alloc arg that we already have. For what I want to achieve in my code, I don't care about the else case.

Reply:
I think the original implementation diverged from the C++ standard, and this change makes it align with the C++ standard when `propagate_on_container_move_assignment` is true.

Thanks though!

Reply:
I think https://en.cppreference.com/w/cpp/named_req/AllocatorAwareContainer:

> Move constructors obtain their instances of allocators by move-constructing from the allocator belonging to the old container.

sounds like the move constructor should always get the allocator from the other container, regardless of `propagate_on_container_move_assignment`, which is only relevant for move assignment, not construction. That's the old behavior I had.

Anyways, it doesn't matter, this doesn't need to conform to the standard.

Reply:
Ah, I see, you are pointing at a different issue, we aren't moving the allocator from the other container. That does seem like a blatant mistake. I wonder if I didn't do that because `auto_allocator` is weird/broken on that case, that allocator needs to specify *not* to move the allocator. Now that this respects that flag, maybe this can be cleaned up further. I'll take another look at this after your change is merged.